modules/ocf_www/files/build-vhosts

```diff
@@ -16,188 +13,357 @@ import jinja2
+
+    for domain, vhost in get_app_vhosts().items():
+        user = vhost['username']
+
+        # Warn about ocfdev group
+        if 'ocfdev' not in groups_for_user(user):
+            report(
+                "Warning: user '{}' not in group ocfdev but has app vhost '{}'"
+                .format(user, domain)
+            )
+
+        # Fix socket dir
+        socket_dir = str(APP_DIR / user)
+        try:
+            os.makedirs(socket_dir, 0o755)
+        except FileExistsError:
+            pass
+        shutil.chown(socket_dir, user, 'ocf')
+        os.chmod(socket_dir, 0o755)
+
+        # Regenerate SSL bundle if older than cert
+        ssl = SSL(domain)
+        if 'nossl' not in vhost['flags'] and ssl.is_valid and (
+            not os.path.exists(ssl.bundle) or
+            os.path.getmtime(ssl.bundle) < os.path.getmtime(ssl.cert)
+        ):
+            report('Generating SSL bundle for ' + domain)
+            with open(ssl.bundle, 'w') as f, open(ssl.cert) as f1, open(ssl.chain) as f2:
+                shutil.copyfileobj(f1, f)
+                shutil.copyfileobj(f2, f)
+
+
+def main():
     parser = argparse.ArgumentParser(
         description=__doc__,
         formatter_class=argparse.RawTextHelpFormatter,
     )
     parser.add_argument(
+        'target', metavar='TARGET', choices=('web', 'app'),
+        help="One of 'web' or 'app'. Determines which server to build for.",
+    )
+    parser.add_argument(
         '--dry-run', action='store_true',
-        help='Build vhost config file and print its contents without touching Apache.',
+        help=(
+            'Build the vhost config file and print its contents without '
+            'touching any files or services.'
+        ),
     )
     parser.add_argument(
         '--no-reload', action='store_true',
         help=(
-            "Build vhost config file, put it in place, but don't reload Apache. "
-            'Keep in mind that Apache may get reloaded at any time later on.',
+            "Build the vhost config file, put it in place, but don't reload "
+            'the web server. '
+            'Keep in mind that it may get reloaded at any time later on.'
         ),
     )
-    args = parser.parse_args(argv)
+    parser.add_argument(
+        '--dev', action='store_true',
+        help=(
+            'Use development server values, '
+            'i.e. use dev-www/dev-apphost instead of www/apphost.'
+        ),
+    )
+    args = parser.parse_args(sys.argv[1:])
 
-    config_file = build_config_file()
+    global target
+    target = args.target
 
-    if args.dry_run:
-        report(config_file)
-        return 0
+    if target == 'app' and not args.dry_run:
+        process_app_vhosts()
 
-    fh, path = tempfile.mkstemp(dir='/etc/apache2', prefix='ocf-vhost.conf')
-    try:
-        os.chmod(path, 0o644)
-        os.write(fh, config_file.encode('utf8'))
-        os.close(fh)
-        report(path)
+    global dev_config
```

Comment:
It could, but as a general rule it kinda sucks to pass around global behavior options (`--debug`, `--verbose`) everywhere you go and insert them into your data types. :(

In this case it's pretty easy to do, though, so I may as well.